openedx_external_enrollments/external_enrollments/viper_external_enrollment.py

```diff
@@ -0,0 +1,133 @@
+        return 'viper'
+
+    def _execute_post(self, url, data=None, headers=None, json_data=None):
+        """
+        Execute post request.
+        """
+        course_shell_id = json_data.get('variables').pop('shellCourseId')
+        response = requests.post(
+            url=url,
+            json=json_data,
+            headers=headers,
+        )
+        enrollment_data = json_data.get('variables')
+
+        if response.status_code == status.HTTP_200_OK:
+            ExternalEnrollment.objects.update_or_create(  # pylint: disable=no-member
+                controller_name=str(self),
+                course_shell_id=course_shell_id,
+                email=enrollment_data.get('email'),
+                defaults={
+                    'meta': {
+                        'course_id': enrollment_data.get('courseId'),
+                        'class_id': enrollment_data.get('classId'),
+                        'course_url': response.json().get('link'),
+                    },
+                }
+            )
+
+        return response
+
+    def _get_enrollment_headers(self):
+        """
+        Method that returns the required header.
+        """
+        return {
+            'x-api-key': settings.OEE_VIPER_MUTATIONS_API_KEY,
+        }
+
+    def _get_enrollment_data(self, data, course_settings):
+        """
+        Method that provide the relevant data for the enrollment.
+        """
+
+        return {
+            'action': self.CREATE_ENROLLMENT_ACTION,
+            'variables': {
+                'shellCourseId': data.get('course_id'),
+                'classId': course_settings.get('external_course_class_id'),
+                'courseId': course_settings.get('external_course_run_id'),
+                'email': data.get('user_email'),
+                'name': data.get('name'),
+                'provider': settings.OEE_VIPER_IDP,
+            },
+        }
+
+    def _get_enrollment_url(self, course_settings):
+        """
+        Method that returns base url for the Viper integration API.
+        """
+        return settings.OEE_VIPER_API_URL
+
+    def _get_course_home_url(self, course_settings, data=None):
+        """
+        Get course url home.
+        """
+        user = data.get('user')
+
+        try:
+            enrollment = ExternalEnrollment.objects.get(  # pylint: disable=no-member
+                controller_name=str(self),
+                course_shell_id=data.get('course_id'),
+                email=user.email
+            )
+        except ObjectDoesNotExist:
+            LOG.error('Failed to retrieve enrollment for viper, Reason: Does not exist')
+
+            return None
+
+        if enrollment.meta.get('is_active'):
+            return enrollment.meta.get('course_url')
+
+        # Retrieve the invitation/course url through an enrollment.
+        # This operation will happen until invitation link is replaced by the actual course enrollment target.
+        # import pudb;pu.db
```

Comment:
Please remove this line @ivanvgh

Reply:
Is it really necessary to call this API each time a user wants to get "course_home_url" ???

Reply:
It will be called only if the enrollment invitation URL is visited and accepted, when accepted means that the enrollment in viper is created at that point, otherwise the enrollment does not exist yet.
Once the enrollment is accepted, the next time that the learner hit the course card to see the content, this API will be called as last time in order to retrieve the new course launch URL and set the `is_active` flag to `True`. The following hits course launch URL will be retrieved from the  `ExternalEnrollment` object.